state-chain/pallets/cf-elections/src/tests.rs

```diff
@@ -1,91 +1,251 @@
 #![cfg(test)]
-
-use super::*;
-use sp_std::collections::btree_set::BTreeSet;
 use std::collections::BTreeMap;
 
-use cf_traits::EpochInfo;
+use crate::{mock::*, *};
+use cf_primitives::AuthorityCount;
+use electoral_system::{
+	AuthorityVoteOf, ConsensusStatus, ElectionReadAccess, ElectoralReadAccess, ElectoralWriteAccess,
+};
+use electoral_systems::mock::MockElectoralSystem;
+use frame_support::traits::OriginTrait;
+use vote_storage::AuthorityVote;
 
-use electoral_system::{AuthorityVoteOf, ElectoralReadAccess};
-use frame_system::RawOrigin;
-use mock::{new_test_ext, MockEpochInfo, Test, INITIAL_UNSYNCED_STATE};
+pub trait ElectoralSystemTestExt: Sized {
+	fn assume_consensus(self) -> Self;
+	fn assume_no_consensus(self) -> Self;
+	fn expect_consensus(self, expected: ConsensusStatus<AuthorityCount>) -> Self;
+	fn new_election(self) -> Self;
+	fn submit_simple<I: 'static>(
+		self,
+		validator_ids: &[u64],
+		call: impl Fn(&u64) -> Call<Test, I>,
+	) -> Self
+	where
+		Test: Config<I>;
+	fn submit_votes<I: 'static>(
+		self,
+		validator_ids: &[u64],
+		vote: AuthorityVoteOf<MockElectoralSystem>,
+		expected_error: Option<Error<Test, I>>,
+	) -> Self
+	where
+		Test: Config<I, ElectoralSystem = MockElectoralSystem>;
+}
 
-#[test]
-fn happy_path_vote_and_consensus() {
-	const NEW_DATA: u64 = 23;
-
-	fn submit_vote_for(validator_id: u64) {
-		let electoral_data = Pallet::<Test, Instance1>::electoral_data(&validator_id).unwrap();
-
-		for (election_identifier, _election_data) in electoral_data.current_elections {
-			Pallet::<Test, Instance1>::vote(
-				RawOrigin::Signed(validator_id).into(),
-				BoundedBTreeMap::try_from(
-					[(
-						election_identifier,
-						AuthorityVoteOf::<<Test as Config<Instance1>>::ElectoralSystem>::Vote(
-							NEW_DATA,
-						),
-					)]
-					.into_iter()
-					.collect::<BTreeMap<_, _>>(),
-				)
-				.unwrap(),
-			)
-			.unwrap();
-		}
+impl ElectoralSystemTestExt for TestRunner<TestContext> {
+	/// Starts a new election, adding its unique monotonic identifier to the test context.
+	#[track_caller]
+	fn new_election(self) -> Self {
+		self.then_execute_with(
+			#[track_caller]
+			|mut ctx| {
+				let unique_monotonic_identifier =
+					*Pallet::<Test, Instance1>::with_electoral_access(|electoral_access| {
+						electoral_access.new_election((), (), ())
+					})
+					.expect("New election should not corrupt storage.")
+					.election_identifier()
+					.unique_monotonic();
+
+				assert_eq!(Status::<Test, Instance1>::get(), Some(ElectoralSystemStatus::Running));
+
+				Pallet::<Test, Instance1>::with_electoral_access(|electoral_access| {
+					electoral_access
+						.election(ElectionIdentifier::new(unique_monotonic_identifier, ()))
+				})
+				.expect("Expected an initial election.");
+
+				ctx.umis.push(unique_monotonic_identifier);
```

Comment:
In the pallet logic itself, we use ElectionProperties as the source of truth about a particular UMI's existence. See `ensure_election_exists` - we could use ElectionProperties here instead - or at least assert that the UMIs are the same in each - `ElectionState` too - or perhaps we call `ensure_election_exixsts` here to avoid getting too close to implementation details

Reply:
Yes, I wanted to avoid calling into any storage items directly. For example if we rely on ElectionProperties to get the list of active elections, then `ensure_election_exists` will always trivially pass.

Also, this `umis` item is not intended as a list of *live* elections - it could also be used to eg. assert that an election has been deleted. It's really just a record of what elections have been created during a particular test run.

I'm not sure there's much value in adding an assert here that the election exists: we just created it, and `ensure_election_exists` (which is a private method) is implicitly called in many places already.

Reply:
As discussed - I added an integrity test to check the storage invariant.